README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# coviduk

<!-- badges: start -->

[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
<img src='man/figures/logo.png' align="right" height="139" />
<!-- badges: end -->

The goal of coviduk is to make it easier to get data out of the official
[UK SARS-CoV-2 dashboard](https://coronavirus.data.gov.uk/). It does
this by wrapping up the API calls in simpler functions, reducing the
need to set various parameters manually and reduce repetitive code.

Currently the package only handles hospital admissions data (new daily
admissions and cumulative totals) and case numbers (by reported and
specimen date), but a more expansive range of data will be built in, at
least until the government’s developers release the R package promised
in their [developer’s
guide](https://coronavirus.data.gov.uk/developers-guide).

## Installation

You can install the development version of coviduk from Github.

``` r
devtools::install_github("shanej90/coviduk", ref = "main")
```

## How it works

There are three in use functions now. Brief details and examples below,
but there is a [reference guide](https://shanej90.github.io/coviduk/).

`get_hospital_admissions()` will call hospitilisation data for a chosen
region and date range. The results will be presented as a dataframe,
stating the date, region and number of new and cumulative/total
admissions since records started.

The parameters you input will create a query to send to the coronavirus
dashboard API, saving you from having to type out the whole query. For
regions you can select any of the four UK nations, or the English NHS
regions.

`get_case_data()` draws on data of confirmed cases, both by reporting
date and date the test specimen was taken. Note these two variables are
available at different ‘levels’ of UK areas (so what is available for
natons may not be available for local authorities and vice versa).

Note that you can/will be throttled if you make too many requests -
you’ll get an error message explaining this if so. As such, better to
keep the number of requests low and/or leave gaps to avoid being
throttled.

`get_deaths_data()` draws data on deaths within 28 days of a positive
test, by both date reported and date of death. I intend to bring in the
weekly Office for National Statistics data on death certificates
mentioning the disease, which I’d take as a truer measure (although they
are only by date of registration rather than actual death,
unfortunately).

## Example

This is a basic example which shows how you can get hold of the hospital
admissions data for a specific timeframe.

``` r
library(coviduk)

example <- coviduk::get_hospital_admissions("London", "2020-09-01", "2020-09-15")

head(example)
#>         date area_name area_code new_admissions total_admissions
#> 1 2020-09-01    London E40000003              6            24809
#> 2 2020-09-02    London E40000003             11            24820
#> 3 2020-09-03    London E40000003             15            24835
#> 4 2020-09-04    London E40000003             11            24846
#> 5 2020-09-05    London E40000003             23            24869
#> 6 2020-09-06    London E40000003             12            24881
```

And this example shows how you can get case data - also highlighting
that some data isn’t available at all available area structure levels.
Deaths data will be very similar but obviously with deaths, instead of
case, data.

``` r
library(coviduk)

example2 <- coviduk::get_case_data("ltla", "Exeter", "2020-09-01", "2020-09-15")

head(example2)
#>         date area_code area_name reported_new reported_total specimen_new
#> 1 2020-09-01 E07000041    Exeter            0             NA            1
#> 2 2020-09-02 E07000041    Exeter            0             NA            2
#> 3 2020-09-03 E07000041    Exeter            1             NA            5
#> 4 2020-09-04 E07000041    Exeter            0             NA            5
#> 5 2020-09-05 E07000041    Exeter            3             NA            1
#> 6 2020-09-06 E07000041    Exeter            2             NA            2
#>   specimen_total
#> 1            275
#> 2            277
#> 3            282
#> 4            287
#> 5            288
#> 6            290
```


NEWS.md
# coviduk 0.3.0
* Added new function - get deaths data by either reporting date or date of death.

# coviduk 0.2.0

* Added new function - get case data by reported and specimen date.
* Updated functions to pull area code alongside other data.

# coviduk 0.1.2

* Added ability to pull UK nations data.

# covid 0.1.1

* Improved error handling and update of documentation to highlight potential 429 responses (too many requests).

# covid 0.1.0

* Launched with single function to pull hospital admissions for English NHS regions.


tests/testthat.R
library(testthat)
library(coviduk)

test_check("coviduk")


tests/testthat/test-get_case_data.R
#test error messages come back correctly----------------

#regions
testthat::test_that(
  "region outside list throws error",
  expect_error(
    coviduk::get_case_data("nation", "hello", "2020-05-20", "2020-06-15"),
    "nation must be one of: England; Northern Ireland; Scotland; Wales",
    fixed = T
    )
  )

#format--------------------------------

#make sure you end up with a dataframe
testthat::expect_is(coviduk::get_case_data("ltla", "Exeter", "2020-06-30", "2020-08-01"), "data.frame")


tests/testthat/test-get_deaths_data.R
#test error messages come back correctly----------------

#regions
testthat::test_that(
  "region outside list throws error",
  expect_error(
    coviduk::get_deaths_data("nation", "hello", "2020-05-20", "2020-06-15"),
    "nation must be one of: England; Northern Ireland; Scotland; Wales",
    fixed = T
    )
  )

#format--------------------------------

#make sure you end up with a dataframe
testthat::expect_is(coviduk::get_deaths_data("ltla", "Exeter", "2020-06-30", "2020-08-01"), "data.frame")


R/get_deaths_data.R
#' Create a dataframe of deaths data, according to the specifications you set.
#'
#' Makes API call to UK Government coronavirus dashboard and collects data on deaths according to the parameters you set. Data will be by nation, region or local authority (upper or lower tier). Note only 1,000 results max returned and that throttling can be implemented if you make too many requests.
#'
#' @param area_type Specify the name of the area for which you want to get the data.
#' @param area_name Specify the area (by name, not code) for which you would like to pull the data.
#' @param start_date Specify the first date for which you wish to collect the data. ymd format.
#' @param end_date Specify the last date for which you wish to collect the data. ymd format.
#' @return A dataframe containing deaths data by the specified area type, by date reported and of death. Note not all variables are available for all area types.
#' @export

get_deaths_data <- function(area_type, area_name, start_date, end_date) {

  #area data---------------

  #local authorities
  las <- read.csv("http://geoportal1-ons.opendata.arcgis.com/datasets/3e4f4af826d343349c13fb7f0aa2a307_0.csv")

  #nations
  nations <- c("England", "Northern Ireland", "Scotland", "Wales")

  #regions
  regions <- c("East of England", "London", "East Midlands", "West Midlands", "North East", "Yorkshire and The Humber", "North West", "South East", "South West")


  #error handling----------------------------------

  #area type
  if(!area_type %in% c("nation", "region", "utla", "ltla")) stop("area_type must be one of: nation; region; utla; ltla")

  #valid areas
  if(area_type == "nation" & !area_name %in% nations) stop("nation must be one of: England; Northern Ireland; Scotland; Wales")
  if(area_type == "region" & !area_name %in% regions) stop("region must be one of: East Midlands; East of England; North East; North West; South East; South West; West Midlands; Yorkshire and The Humber")
  if(area_type == "utla" & !area_name %in% las$UTLA19NM) stop("utla must be as per http://geoportal1-ons.opendata.arcgis.com/datasets/3e4f4af826d343349c13fb7f0aa2a307_0.csv")
  if(area_type == "ltla" & !area_name %in% las$LTLA19NM) stop("ltla must be as per http://geoportal1-ons.opendata.arcgis.com/datasets/3e4f4af826d343349c13fb7f0aa2a307_0.csv")

  #make sure end date greater than or equal to start date
  if(lubridate::ymd(end_date) < lubridate::ymd(start_date)) stop("end date must be later than, or same as, start date")

  #make sure start date no earlier than 2020-01-30
  if(lubridate::ymd(start_date) < lubridate::ymd("2020-01-30")) stop("Minimum start date is 2020-01-30")

  #sequence of dates to call-----------------------------------------
  dates_to_call <- seq.Date(lubridate::ymd(start_date), lubridate::ymd(end_date), by = "days") %>%
    format(., "%Y-%m-%d")

  #create query parameters-------------------------------------

  call_api <- function(.date) {

    #filter settings
    filter_settings <- c(
      glue::glue("areaType={area_type}"),
      glue::glue("areaName={area_name}"),
      glue::glue("date={.date}")
    )

    #structure settings
    structure_settings <- list(
      date = .date,
      area = list(code =  "areaCode", name = "areaName"),
      reported = list(new = "newDeaths28DaysByPublishDate", total = "cumDeaths28DaysByPublishDate"),
      of_death = list(new = "newDeaths28DaysByDeathDate", total = "cumDeaths28DaysByDeathDate")
    )

    #query-------------------------------------

    #base url
    endpoint <- "https://api.coronavirus.data.gov.uk/v1/data"

    #TODO: when polite package containing politely function released on CRAN, do things politely
    #polite_get <- polite::politely(httr::GET, verbose = T)

    #call api and convert to appropriate format
    result <- httr::GET(
      url = endpoint,
      #convert queries into JSON format
      query = list(
        filters = paste(filter_settings, collapse = ";"),
        structure = jsonlite::toJSON(structure_settings, auto_unbox = T)
      ),
      httr::timeout(10)
    )

    if(result$status_code >= 400) {
      err_msg = httr::http_status(result)
      stop(err_msg)
    }

    #result text
    result_text <- httr::content(result, "text")

    #turn into dataframe
    jsonlite::fromJSON(result_text)[["data"]] %>%
      #sort out listed nature of dataframe..probably a better way of doing this
      dplyr::mutate(
        area_code = area$code,
        area_name = area$name,
        reported_new = reported$new,
        reported_total = reported$total,
        of_death_new = of_death$new,
        of_death_total = of_death$total
      ) %>%
      #remove unnecessary columns
      dplyr::select(-reported, -of_death, -area) %>%
      #set sensible names
      #set sensible names
      stats::setNames(c("date", "area_code", "area_name", "reported_new", "reported_total", "of_death_new", "of_death_total"))

  }

  #run function for all specified dates, and bind rows
  lapply(dates_to_call, call_api) %>%
    dplyr::bind_rows(.)

}


R/get_hospital_admissions.R
#' Create a dataframe of hospital admissions data, according to the specifications you set.
#'
#' Makes API call to UK Government coronavirus dashboard and collects data on admissions according to the parameters you set. Data will be by NHS region or nation. Note only 1,000 results max returned and that throttling can be implemented if you make too many requests.
#'
#'
#' @param area_name Specify the region/nation (by name, not code) for which you want to get the data.
#' @param start_date Specify the first date for which you wish to collect the data. ymd format.
#' @param end_date Specify the last date for which you wish to collect the data. ymd format.
#' @return A dataframe containing hospital admissions data by NHS region or nation: new admissions and cumulative admissions. Only a maximum of 1,000 results will be returned.
#' @export

get_hospital_admissions <- function(area_name, start_date, end_date) {

  #error checking/validation-----------------------------------------

  #valid regions
  valid_regions <- c("East of England", "London", "Midlands", "North East and Yorkshire", "North West", "South East", "South West", "Scotland", "Wales", "Northern Ireland", "England")

  #make sure valid NHS region is chosen
  if(!area_name %in% valid_regions) stop(glue::glue("`area` must be on of: {paste(valid_regions, collapse = ';')}"))

  #make sure dates are in right format
  if(
    stringr::str_detect(start_date, "[0-9][0-9][0-9][0-9]-[0-1][0-9]-[0-3][0-9]") == F |
    stringr::str_detect(end_date, "[0-9][0-9][0-9][0-9]-[0-1][0-9]-[0-3][0-9]") == F
  ) stop("Dates must be in 'yyyy-mm-dd' format.")

  #make sure end date greater than or equal to start date
  if(lubridate::ymd(end_date) < lubridate::ymd(start_date)) stop("end date must be later than, or same as, start date")

  #make sure start date no earlier than 2020-03-19
  if(lubridate::ymd(start_date) < lubridate::ymd("2020-03-19")) stop("Minimum start date is 2020-03-19")

  #sequence of dates to call-----------------------------------------
  dates_to_call <- seq.Date(lubridate::ymd(start_date), lubridate::ymd(end_date), by = "days") %>%
    format(., "%Y-%m-%d")

  #create query parameters-------------------------------------------------

  #base url
  endpoint <- "https://api.coronavirus.data.gov.uk/v1/data"

  #area type
  area_type <- ifelse(area_name %in% c("England", "Wales", "Northern Ireland", "Scotland"), "nation", "nhsRegion")

  #function to run the query for each registered date--------------------------
  call_api <- function(.date) {

    #filter settings
    filter_settings <- c(
      glue::glue("areaType={area_type}"),
      glue::glue("areaName={area_name}"),
      glue::glue("date={.date}")
      )

    #structure settings
    structure_settings <- list(
      date = .date,
      area = list(name = "areaName", code = "areaCode"),
      hospital_admissions = list(new = "newAdmissions", total = "cumAdmissions")
      )

    #TODO: when polite package containing politely function released on CRAN, do things politely
    #polite_get <- polite::politely(httr::GET, verbose = T)

    #call api and convert to appropriate format
    result <- httr::GET(
      url = endpoint,
      #convert queries into JSON format
      query = list(
        filters = paste(filter_settings, collapse = ";"),
        structure = jsonlite::toJSON(structure_settings, auto_unbox = T)
        ),
      httr::timeout(10)
    )

    if(result$status_code >= 400) {
      err_msg = httr::http_status(result)
      stop(err_msg)
    }

    #result text
    result_text <- httr::content(result, "text")

    #turn into dataframe
      jsonlite::fromJSON(result_text)[["data"]] %>%
      #sort out listed nature of dataframe..probably a better way of doing this
      dplyr::mutate(
        area_name = area$name,
        area_code = area$code,
        new = hospital_admissions$new,
        total = hospital_admissions$total
        ) %>%
      #remove unnecessary columns
      dplyr::select(-area, -hospital_admissions) %>%
      #set sensible names
      stats::setNames(c("date", "area_name", "area_code", "new_admissions", "total_admissions"))

      }

  #run function for all specified dates, and bind rows
  lapply(dates_to_call, call_api) %>%
    dplyr::bind_rows(.)

}


R/get_case_data.R
#' Create a dataframe of case data, according to the specifications you set.
#'
#' Makes API call to UK Government coronavirus dashboard and collects data on cases according to the parameters you set. Data will be by nation, region or local authority (upper or lower tier). Note only 1,000 results max returned and that throttling can be implemented if you make too many requests.
#'
#' @param area_type Specify the name of the area for which you want to get the data.
#' @param area_name Specify the area (by name, not code) for which you would like to pull the data.
#' @param start_date Specify the first date for which you wish to collect the data. ymd format.
#' @param end_date Specify the last date for which you wish to collect the data. ymd format.
#' @return A dataframe containing case data by the specified area type, by date reported and of specimen. Note not all variables are available for all area types.
#' @export

get_case_data <- function(area_type, area_name, start_date, end_date) {

  #area data---------------

  #local authorities
  las <- read.csv("http://geoportal1-ons.opendata.arcgis.com/datasets/3e4f4af826d343349c13fb7f0aa2a307_0.csv")

  #nations
  nations <- c("England", "Northern Ireland", "Scotland", "Wales")

  #regions
  regions <- c("East of England", "London", "East Midlands", "West Midlands", "North East", "Yorkshire and The Humber", "North West", "South East", "South West")


  #error handling----------------------------------

  #area type
  if(!area_type %in% c("nation", "region", "utla", "ltla")) stop("area_type must be one of: nation; region; utla; ltla")

  #valid areas
  if(area_type == "nation" & !area_name %in% nations) stop("nation must be one of: England; Northern Ireland; Scotland; Wales")
  if(area_type == "region" & !area_name %in% regions) stop("region must be one of: East Midlands; East of England; North East; North West; South East; South West; West Midlands; Yorkshire and The Humber")
  if(area_type == "utla" & !area_name %in% las$UTLA19NM) stop("utla must be as per http://geoportal1-ons.opendata.arcgis.com/datasets/3e4f4af826d343349c13fb7f0aa2a307_0.csv")
  if(area_type == "ltla" & !area_name %in% las$LTLA19NM) stop("ltla must be as per http://geoportal1-ons.opendata.arcgis.com/datasets/3e4f4af826d343349c13fb7f0aa2a307_0.csv")

  #make sure end date greater than or equal to start date
  if(lubridate::ymd(end_date) < lubridate::ymd(start_date)) stop("end date must be later than, or same as, start date")

  #make sure start date no earlier than 2020-01-30
  if(lubridate::ymd(start_date) < lubridate::ymd("2020-01-30")) stop("Minimum start date is 2020-01-30")

  #sequence of dates to call-----------------------------------------
  dates_to_call <- seq.Date(lubridate::ymd(start_date), lubridate::ymd(end_date), by = "days") %>%
    format(., "%Y-%m-%d")

  #create query parameters-------------------------------------

  call_api <- function(.date) {

    #filter settings
    filter_settings <- c(
      glue::glue("areaType={area_type}"),
      glue::glue("areaName={area_name}"),
      glue::glue("date={.date}")
      )

    #structure settings
    structure_settings <- list(
      date = .date,
      area = list(code =  "areaCode", name = "areaName"),
      reported = list(new = "newCasesByPublishDate", total = "cumCasesByPublishDate"),
      specimen = list(new = "newCasesBySpecimenDate", total = "cumCasesBySpecimenDate")
      )

    #query-------------------------------------

    #base url
    endpoint <- "https://api.coronavirus.data.gov.uk/v1/data"

    #TODO: when polite package containing politely function released on CRAN, do things politely
    #polite_get <- polite::politely(httr::GET, verbose = T)

    #call api and convert to appropriate format
    result <- httr::GET(
      url = endpoint,
      #convert queries into JSON format
      query = list(
        filters = paste(filter_settings, collapse = ";"),
        structure = jsonlite::toJSON(structure_settings, auto_unbox = T)
        ),
      httr::timeout(10)
      )

    if(result$status_code >= 400) {
      err_msg = httr::http_status(result)
      stop(err_msg)
      }

    #result text
    result_text <- httr::content(result, "text")

    #turn into dataframe
    jsonlite::fromJSON(result_text)[["data"]] %>%
      #sort out listed nature of dataframe..probably a better way of doing this
      dplyr::mutate(
        area_code = area$code,
        area_name = area$name,
        reported_new = reported$new,
        reported_total = reported$total,
        specimen_new = specimen$new,
        specimen_total = specimen$total
        ) %>%
      #remove unnecessary columns
      dplyr::select(-reported, -specimen, -area) %>%
      #set sensible names
      stats::setNames(c("date", "area_code", "area_name", "reported_new", "reported_total", "specimen_new", "specimen_total"))

    }

  #run function for all specified dates, and bind rows
  lapply(dates_to_call, call_api) %>%
    dplyr::bind_rows(.)

}
